lib/Cro/Iri.pm6

```diff
@@ -1,75 +1,253 @@
+            %parts<query> = $<iquery>.ast if $<iquery>;
+            %parts<fragment> = $<ifragment>.ast if $<ifragment>;
+            make Cro::Iri.bless(|%parts);
+        }
+
+        method ihier-part:sym<authority>($/) {
+            make {
+                path => $<ipath-abempty>.ast,
+                $<iauthority>.ast
+            }
+        }
+
+        method ihier-part:sym<absolute>($/) {
+            make (path => $<ipath-absolute>.ast);
+        }
+
+        method ihier-part:sym<rootless>($/) {
+            make (path => $<ipath-rootless>.ast);
+        }
+
+        method ihier-part:sym<empty>($/) {
+            make (path => $<ipath-empty>.ast);
+        }
+
+        method iauthority($/) {
+            make {
+                authority => ~$/,
+                port => $<port> ?? $<port>.ast !! Int,
+                userinfo => $<iuserinfo> ?? ~$<iuserinfo> !! Str,
+                $<ihost>.ast
+            };
+        }
+
+        method ihost:sym<IPv4address>($/) {
+            make {
+                host => ~$/,
+                host-class => Cro::ResourceIdentifier::Host::IPv4
+            }
+        }
+
+        method ihost:sym<IPv6address>($/) {
+            make {
+                host => ~$/,
+                host-class => Cro::ResourceIdentifier::Host::IPv6
+            }
+        }
+
+        method ihost:sym<IPvFuture>($/) {
+            make {
+                host => ~$/,
+                host-class => Cro::ResourceIdentifier::Host::IPvFuture
+            }
+        }
+
+        method ihost:sym<ireg-name>($/) {
+            make {
+                host => decode-percents(~$/),
+                host-class => Cro::ResourceIdentifier::Host::RegName
+            }
+        }
+
+        method ipath-abempty($/) {
+            make ~$/;
+        }
+
+        method ipath-absolute($/) {
+            make ~$/;
+        }
+
+        method ipath-rootless($/) {
+            make ~$/;
+        }
+
+        method ipath-empty($/) {
+            make '';
+        }
+
+        method iquery($/) {
+            make ~$/;
+        }
+
+        method ifragment($/) {
+            make ~$/;
+        }
+
+        method ref($/) {
+            make ($<IRI> // $<relative-ref>).ast
+        }
+
+        method relative-ref($/) {
+            my %parts = $<irelative-part>.ast;
+            %parts<query> = .ast with $<iquery>;
+            %parts<fragment> = .ast with $<ifragment>;
+            make Cro::Iri.bless(|%parts);
+        }
+    }
+
+    method parse(Str() $iri-string, :$grammar = Cro::Iri::GenericParser,
+                 :$actions = Cro::Iri::GenericActions.new --> Cro::Iri) {
+        with $grammar.parse($iri-string, :$actions) {
+            .ast
+        } else {
+            die X::Cro::Iri::ParseError.new(:$iri-string)
+        }
+    }
+
+    sub encode-percents-except-ASCII(Str $s) {
+        $s.subst: :g, /<-[A..Za..z0..9_.~:/%=-]>+/, {
+            .Str.encode('utf8').list.map({ sprintf '%%%02s', .base(16) }).join
```

Comment:
I suspect the cost of `sprintf` is huge in this context, and we'd be better just checking if the `base` call gives 1 char back and padding it with 0 ourselves.